prbt_hardware_support/src/brake_test_executor.cpp

```diff
@@ -15,92 +15,85 @@
+  : detect_robot_motion_func_(detect_robot_motion_func)
+  , hold_controller_func_(controller_hold_func)
+  , execute_brake_test_func_(trigger_brake_test_func)
+  , unhold_controller_func_(unhold_func)
+  , brake_test_result_func_(brake_test_result_func)
 {
-  brake_test_srv_ = nh_.advertiseService(EXECUTE_BRAKETEST_SERVICE_NAME,
-                                         &BrakeTestExecutor::executeBrakeTest,
-                                         this);
-
-  waitForService(BRAKETEST_ADAPTER_SERVICE_NAME);
-  trigger_braketest_client_ = nh_.serviceClient<BrakeTest>(BRAKETEST_ADAPTER_SERVICE_NAME);
-
-  waitForService(CONTROLLER_HOLD_MODE_SERVICE_NAME);
-  controller_hold_client_ = nh_.serviceClient<std_srvs::Trigger>(CONTROLLER_HOLD_MODE_SERVICE_NAME);
+  if (!detect_robot_motion_func_)
+  {
+    throw BrakeTestExecutorException("Function to detect robot motion missing");
+  }
 
-  waitForService(CONTROLLER_UNHOLD_MODE_SERVICE_NAME);
-  controller_unhold_client_ = nh_.serviceClient<std_srvs::Trigger>(CONTROLLER_UNHOLD_MODE_SERVICE_NAME);
+  if (!execute_brake_test_func_)
+  {
+    throw BrakeTestExecutorException("Function to trigger brake test on robot missing");
+  }
 
-  waitForService(BRAKE_TEST_RESULT_SERVICE_NAME);
-  brake_test_result_client_ = nh_.serviceClient<SendBrakeTestResult>(BRAKE_TEST_RESULT_SERVICE_NAME);
+  if (!brake_test_result_func_)
+  {
+    throw BrakeTestExecutorException("Function to send brake test result to FS controller missing");
+  }
 }
 
 bool BrakeTestExecutor::executeBrakeTest(BrakeTest::Request& /*req*/,
-                                         BrakeTest::Response& response)
+                                         BrakeTest::Response& res)
 {
-  if (BrakeTestUtils::detectRobotMotion())
+  if (detect_robot_motion_func_())
   {
-    response.success = false;
-    response.error_msg = "Robot is moving, cannot perform brake test";
-    response.error_code.value = BrakeTestErrorCodes::ROBOT_MOTION_DETECTED;
+    res.success = false;
+    res.error_msg = "Robot is moving, cannot perform brake test";
+    res.error_code.value = BrakeTestErrorCodes::ROBOT_MOTION_DETECTED;
     return true;
   }
 
   ROS_INFO_STREAM("Enter hold for braketest. Do not unhold the controller!");
 
-  std_srvs::Trigger trigger_srv;
-  if (!controller_hold_client_.call(trigger_srv))
+  if (hold_controller_func_)
   {
-    ROS_WARN_STREAM("Failed to trigger hold via service " << controller_hold_client_.getService());
+    hold_controller_func_();
   }
-
-  BrakeTest braketest_srv;
-  if (!trigger_braketest_client_.call(braketest_srv))
+  else
   {
-    response.success = false;
-    response.error_msg = "Failed to trigger brake test via service " + trigger_braketest_client_.getService();
-    response.error_code.value = BrakeTestErrorCodes::TRIGGER_BRAKETEST_SERVICE_FAILURE;
+    ROS_WARN_STREAM("Calling hold on controller skipped because no hold service given");
```

Comment:
We need better error handling. If a hold is not possible the whole system is in a critical state...

Reply:
There is always true returned...

Reply:
And this DOES change behaviour (hold was required before).

Reply:
> There is always true returned...

Services always return true, if the result type was filled, see [ROS wiki](http://wiki.ros.org/roscpp/Overview/Services#Callback_Signature).

Reply:
Yes, I agree. Can we do the error handling separately?

Reply:
This PR should not change behaviour. I will correct it.